### PR TITLE
SAA-559 - Wire in prisoner allocated audit event

### PR DESCRIPTION
**Pull request by sioldham**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/audit/PrisonerAllocatedEvent.kt [hunk 1]
```diff
@@ -1,26 +1,22 @@
...
   val activityId: Long,
   val activityName: String,
   val prisonCode: String,
   val prisonerNumber: String,
   val scheduleId: Long,
-  val date: LocalDate,
-  val startTime: LocalTime,
-  val endTime: LocalTime,
+  val scheduleDescription: String,
   createdAt: LocalDateTime,
```
Comment (sioldham) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/audit/PrisonerAllocatedEvent.kt:13]: I added these when I created the events as I guessed they'd be required but they aren't.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableEntityListenerTest.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  private lateinit var auditService: AuditService
+  private val activity = activityEntity()
+  private val allocation = allocation()
+
+  private val activityCreatedCaptor = argumentCaptor<ActivityCreatedEvent>()
+  private val prisonerAllocatedCaptor = argumentCaptor<PrisonerAllocatedEvent>()
+
+  @BeforeEach
+  fun setup() {
+    SecurityTestUtils.setLoggedInUser("Bob")
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableEntityListenerTest.kt:34]: So you can now just annotate any classes that require a logged in user with an extension annotation/class.

```
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
@ActiveProfiles("test")
@ExtendWith(FakeSecurityContext::class)
class AuditableEntityListenerTest(@Autowired private val listener: AuditableListener) {
  ...
}


```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableListener.kt [hunk 3]
```diff
@@ -21,7 +22,14 @@ class AuditableListener {
...
   @PostPersist
-  fun onCreate(activity: Activity) {
+  fun onCreate(entity: Any) {
+    when (entity) {
+      is Activity -> audit(entity)
+      is Allocation -> audit(entity)
+    }
+  }
+
+  private fun audit(activity: Activity) {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableListener.kt:32]: Could get away with one private audit function here and just pass the entity.toXxxxEvent() and any potential failure message to one function e.g.

 ```
@PostPersist
 fun onCreate(entity: Any) {
    when (entity) {
      is Activity -> audit(entity.toActivityCreatedEvent(), "Failed to audit activity creation event for activity id ${entity.activityId}")
      is Allocation -> audit(entity.toPrisonerAllocatedEvent(), "Failed to audit prisoner allocation event for allocation id ${entity.allocationId}")
    }
  }

  private fun audit(event: AuditableEvent, failureMessage: String) {
    runCatching {
      auditing.logEvent(event)
    }.onFailure {
      log.error(failureMessage, it,)
    }
  }

```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt [hunk 4]
```diff
@@ -318,3 +319,14 @@ fun EntityActivity.toActivityCreatedEvent() = ActivityCreatedEvent(
   startDate = startDate,
   createdAt = createdTime,
 )
+
+fun EntityAllocation.toPrisonerAllocatedEvent() = PrisonerAllocatedEvent(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt:323]: Just a thought.  Perhaps we introduce a transformations file just for audit transformations?

**Comment by matt-crowson-moj** [comment]
Looks good.  Just a few comments to think about.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableListener.kt (hunk 3 above)
Comment (sioldham) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableListener.kt:32]: Nice

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableEntityListenerTest.kt (hunk 2 above)
Comment (sioldham) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AuditableEntityListenerTest.kt:34]: Tidy!

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**